pennylane/optimize/rotosolve.py

```diff
@@ -12,172 +12,452 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """Rotosolve gradient free optimizer"""
+# pylint: disable=too-many-branches,cell-var-from-loop
 
 import numpy as np
+from scipy.optimize import brute, shgo
 from pennylane.utils import _flatten, unflatten
 
 
+def _assert_integer(x):
+    x_type = type(x)
+    if not np.issubdtype(x_type, np.integer):
+        raise ValueError(
+            "The numbers of frequencies are expected to be integers. " f"Received {x_type}."
+        )
+
+
+def _validate_num_freqs(num_freqs, requires_grad):
+    if num_freqs is None:
+        num_freqs = [1] * sum(requires_grad)
+        num_freqs_flat = num_freqs
+    elif np.isscalar(num_freqs):
+        _assert_integer(num_freqs)
+        num_freqs = [num_freqs] * sum(requires_grad)
+        num_freqs_flat = num_freqs
+    else:
+        num_freqs_flat = []
+        for num_frequency in num_freqs:
+            if np.isscalar(num_frequency):
+                _assert_integer(num_frequency)
+                num_frequency_flat = num_frequency
+            else:
+                num_frequency_flat = list(_flatten(num_frequency))
+                _ = [_assert_integer(_num_freq) for _num_freq in num_frequency_flat]
```

Comment:
```suggestion
                [_assert_integer(_num_freq) for _num_freq in num_frequency_flat]
```

Reply:
This will make CodeFactor complain... Write this as a full loop?

Reply:
Wait... really? In that case yeah, I would probably rewrite it as a loop!